pkg/testsuite/mock/node.go

```diff
@@ -165,78 +182,12 @@ func (n *Node) hookEvents() {
-	})
-
-	events.Network.AttestationsReceived.Hook(func(commitment *model.Commitment, attestations []*iotago.Attestation, merkleProof *merklehasher.Proof[iotago.Identifier], source peer.ID) {
-		fmt.Printf("%s > Network.AttestationsReceived: from %s %s number of attestations: %d with merkleProof: %s - %s\n", n.Name, source, commitment.ID(), len(attestations), lo.PanicOnErr(merkleProof.JSONEncode()), lo.Map(attestations, func(a *iotago.Attestation) iotago.BlockID {
-			return lo.PanicOnErr(a.BlockID())
-		}))
-	})
-
-	events.Network.AttestationsRequestReceived.Hook(func(id iotago.CommitmentID, source peer.ID) {
-		fmt.Printf("%s > Network.AttestationsRequestReceived: from %s %s\n", n.Name, source, id)
-	})
-
-	events.Network.WarpSyncResponseReceived.Hook(func(id iotago.CommitmentID, ds map[iotago.CommitmentID]iotago.BlockIDs, m *merklehasher.Proof[iotago.Identifier], ds2 iotago.TransactionIDs, m2 *merklehasher.Proof[iotago.Identifier], id2 peer.ID) {
-		fmt.Printf("%s > Network.WarpSyncResponseReceived: from %s %s\n", n.Name, id2, id)
-	})
-
-	events.Network.WarpSyncRequestReceived.Hook(func(id iotago.CommitmentID, id2 peer.ID) {
-		fmt.Printf("%s > Network.WarpSyncRequestReceived: from %s %s\n", n.Name, id2, id)
-	})
-
-	// events.ChainManager.CommitmentBelowRoot.Hook(func(commitmentID iotago.CommitmentID) {
-	//	fmt.Printf("%s > ChainManager.CommitmentBelowRoot: %s\n", n.Name, commitmentID)
-	// })
-
-	events.ChainManager.ForkDetected.Hook(func(fork *chainmanager.Fork) {
-		fmt.Printf("%s > ChainManager.ForkDetected: %s\n", n.Name, fork)
-	})
-
-	// events.Engine.TipManager.BlockAdded.Hook(func(tipMetadata tipmanager.TipMetadata) {
-	//	fmt.Printf("%s > TipManager.BlockAdded: %s in pool %d\n", n.Name, tipMetadata.ID(), tipMetadata.TipPool().Get())
-	// })
-
-	events.CandidateEngineActivated.Hook(func(e *engine.Engine) {
-		fmt.Printf("%s > CandidateEngineActivated: %s, ChainID:%s Slot:%s\n", n.Name, e.Name(), e.ChainID(), e.ChainID().Slot())
-
-		n.attachEngineLogs(failOnBlockFiltered, e)
-	})
-
-	events.MainEngineSwitched.Hook(func(e *engine.Engine) {
-		fmt.Printf("%s > MainEngineSwitched: %s, ChainID:%s Slot:%s\n", n.Name, e.Name(), e.ChainID(), e.ChainID().Slot())
-	})
-
-	events.MainEngineRestarted.Hook(func(e *engine.Engine) {
-		fmt.Printf("%s > MainEngineRestarted: %s, ChainID:%s Slot:%s\n", n.Name, e.Name(), e.ChainID(), e.ChainID().Slot())
-
-		n.attachEngineLogsWithName(failOnBlockFiltered, e, fmt.Sprintf("Main2 - %s", e.Name()[:8]))
-	})
-
-	events.Network.Error.Hook(func(err error, id peer.ID) {
-		fmt.Printf("%s > Network.Error: from %s %s\n", n.Name, id, err)
-	})
-
-	events.Error.Hook(func(err error) {
-		fmt.Printf("%s > Protocol.Error: %s\n", n.Name, err.Error())
+	n.Protocol.Chains.WithElements(func(chain *protocol.Chain) (teardown func()) {
+		return chain.Engine.OnUpdate(func(_ *engine.Engine, newEngine *engine.Engine) {
+			if newEngine != nil {
```

Comment:
When could `newEngine` be nil?

Reply:
When the chain shuts down the engine (for example if it is no longer the heaviest candidate or it gets evicted / shutdown otherwise).